migrations/steps.py

```diff
@@ -0,0 +1,139 @@
+"""
+General purpose migration steps
+"""
+
+import shutil
+import sys
+import tempfile
+from pathlib import Path
+
+
+class MigrationStep:
+    def validate(self):
+        return True
+
+    def snapshot(self, directory):
+        pass
+
+    def run(self):
+        pass
+
+    def revert(self, directory):
+        pass
+
+    def cleanup(self):
+        pass
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: {self.__dict__}>"
+
+    def __str__(self):
+        return self.__repr__()
+
+
+class PathMigrationStep(MigrationStep):
```

Comment:
AIUI this is an 'abstract' migration step, similar to its parent, and not meant to be invoked directly, whereas, for example, Remove or Symlink can be used directly by people implementing migrations in the future? If that's the case maybe we can add some more comments/documentation to this effect